### PR TITLE
GITHUB-9451: Use PHPUnit to list tests

### PR DESCRIPTION
**Pull request by anaelChardan**
**Description (for Contributor and Core Developer)**

We now decided to use PHPUnit to generate the tests we want to execute because by using "*Integration.php" it is very unconvenient, you can't have abstract that is named the same etc... And, moreover what is better than PHPUnit to list PHPUnit tests to launch? :P

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Comment by anaelChardan** [comment]
@skeleton yeah it wasn't perfect and sometimes we had the problem that we launched some tests that didn't exists (like abstract one). And there is nothing better than the tool itself to list tests to launch.

**Comment by skeleton** [comment]
ok, so it could be nice to have this explanation in the description of the PR ;)

**Comment by anaelChardan** [comment]
@skeleton you're totally right, I've forgot my copy paste :P :stvcqjvd:

File: .circleci/find_phpunit.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
```
Comment (jjanvier) [inline .circleci/find_phpunit.sh]: explain please
and exit -1 if you don't have the argument you expected

Comment (arnolanglade) [reply .circleci/find_phpunit.sh]: :+1:

File: .circleci/run_phpunit.sh [hunk 2]
```diff
@@ -6,7 +6,7 @@ TESTFILES=$@
 fail=0
 for TESTFILE in $TESTFILES; do
     echo $TESTFILE
-    docker-compose exec -T fpm ./vendor/bin/phpunit -c app --log-junit var/tests/phpunit/phpunit_$(uuidgen).xml $TESTFILE
+    docker-compose exec -T fpm ./vendor/bin/phpunit -c app --log-junit var/tests/phpunit/phpunit_$(uuidgen).xml --filter $TESTFILE
```
Comment (arnolanglade) [inline .circleci/run_phpunit.sh]: Why do you need --filter ?

File: .circleci/find_phpunit.sh (hunk 1 above)
Comment (anaelChardan) [reply .circleci/find_phpunit.sh]: Good Idea.

File: .circleci/run_phpunit.sh (hunk 2 above)
Comment (anaelChardan) [reply .circleci/run_phpunit.sh]: Because phpunit is only able to launch file and not class without filter.

File: .circleci/find_phpunit.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#PHPUnit 6.5.3 by Sebastian Bergmann and contributors.
+#
+#Available test(s):
+# - Akeneo\Bundle\RuleEngineBundle\tests\integration\ExecuteRuleWithoutPermissionsAppliedIntegration::testRuleExecutionOnAllProducts
+# - tests\integration\Akeneo\Bundle\RuleEngineBundle\Normalizer\Standard\RuleIntegration::testRule
+# - PimEnterprise\Bundle\ApiBundle\tests\EndToEnd\Controller\Asset\CreateAssetIntegration::testCreationOfAnAsset
+# - PimEnterprise\Bundle\ApiBundle\tests\EndToEnd\Controller\Asset\CreateAssetIntegration::testResponseWhenContentIsEmpty
+# - PimEnterprise\Bundle\ApiBundle\tests\EndToEnd\Controller\Asset\CreateAssetIntegration::testResponseWhenValidationFailed
+
+vendor/bin/phpunit -c app --testsuite=$SUITES --list-tests | tail -n +4 | cut -c4- | cut -d ':' -f 1 | sort | uniq | sed -e 's/\\/\\\\/g'
```
Comment (ahocquard) [inline .circleci/find_phpunit.sh]: I think `sed -e 's/\\/\\\\/g'` is useless, as you can avoid it by surrounding your variable this way "$TESTFILES"

File: .circleci/run_phpunit.sh (hunk 2 above)
Comment (ahocquard) [reply .circleci/run_phpunit.sh]: Phpunit accept a classname. You have just to prefix it with "src".
Btw, I don't know how it works, I tried in local what you did and it doesn't work.

Comment (anaelChardan) [reply .circleci/run_phpunit.sh]: Nop, if you give a class namespace, you have to give it through filter, it's in the doc ;)

**Comment by ahocquard** [comment]
There is a drawback witht his method: if the classname is wrong (bad namespace), it will not execute any test.
When no test is executed, the return code is 0...

File: .circleci/run_phpunit.sh (hunk 2 above)
Comment (anaelChardan) [reply .circleci/run_phpunit.sh]: It can fail on your case depending on the behaviour of your terminal :/. I had this issue once.

**Comment by anaelChardan** [comment]
@ahocquard if the namespace is wrong PHPUnit won't find it, so for me no problem on this point.

**Comment by ahocquard** [comment]
That's the problem: the test is no executed as the namespace is wrong. But the return code is 0 so it skips this test silently...

File: .circleci/find_phpunit.sh (hunk 1 above)
Comment (ahocquard) [reply .circleci/find_phpunit.sh]: return code 1, no negative return code in shell

**Comment by anaelChardan** [comment]
The developer is accountable to look at the namespaces and launch it if needed, we get the same problem with tests that don't start with `test` nor annotated by @test. We won't ensure those edge cases imho.

File: .circleci/find_phpunit.sh (hunk 1 above)
Comment (anaelChardan) [reply .circleci/find_phpunit.sh]: Every code that is different from 0 is good for returning a bad execution. But yes 1 is more conventional, I update it.

File: .circleci/run_phpunit.sh (hunk 2 above)
Comment (anaelChardan) [reply .circleci/run_phpunit.sh]: I may work with certain organisation of file

**Comment by ahocquard** [comment]
I don't agree. What about moving tests? Changing autloloading, etc? You have to execute all your tests one after the other?
We are precisely doing this `list-tests` to improve the reliability of the command and to execute tests not ending by "Integration".

FYI, we already had this problem of tests that are not executed, and it will occur very often with big surprise if we don't handle it properly. It should not fail silently.

**Comment by anaelChardan** [comment]
I don't see how we could prevent that. If the tool don't find a namespace, this is due to a misconfiguration problem and it is the accountability of the person who did this refactoring. How could we say "This namespace is missing". The only thing I can see is to look at the number of tests executed before and after. What would you suggest?

File: .circleci/find_phpunit.sh (hunk 3 above)
Comment (anaelChardan) [reply .circleci/find_phpunit.sh]: Phpunit needs double antislash for namespaces :/ : @see https://phpunit.readthedocs.io/en/7.0/textui.html#textui-examples-filter-patterns

File: .circleci/run_phpunit.sh (hunk 2 above)
Comment (ahocquard) [reply .circleci/run_phpunit.sh]: :+1:  for the documentation
I have to surround it with single quote both with bash and zsh.
We should do that, as otherwise it means the command works in circle ci but not in local.
Could you test please?

Comment (anaelChardan) [reply .circleci/run_phpunit.sh]: I've tried but I didn't find a way to add single quotes.

File: .circleci/find_phpunit.sh (hunk 3 above)
Comment (ahocquard) [reply .circleci/find_phpunit.sh]: ok, not without `--filter` for your information ( I tested). Weird behavior

**Comment by anaelChardan** [comment]
So, according to some tests using wrong namespaces PHPUnit does find them so the problem doesn't exist. We should have a script on the CI to find all bad namespaces.

A script like that can do that:

```php
<?php

$loader = require __DIR__.'/../app/autoload.php';

function phpUnitNamespaces() {
    $outputs = [];
    exec("vendor/bin/phpunit -c app --testsuite=PIM_Integration_Test --list-tests | tail -n +4 | cut -c4- | cut -d ':' -f 1 | sort | uniq", $outputs);

    $namespaces = [];
    foreach ($outputs as $output) {
        $namespaces[] = [
            'namespace' => $output,
            'file' => (new \ReflectionClass($output))->getFileName()
        ];
    }
    return $namespaces;
}



var_dump(phpUnitNamespaces());
```

File: .circleci/find_phpunit.sh (hunk 3 above)
Comment (anaelChardan) [reply .circleci/find_phpunit.sh]: Yes, it's weird, their not consistent
